File: 08FlowControl/match/destructure_structures.cs
using  System;
public struct Foo {
    public (int, int) X { get; }
    public int Y { get; }

    public Foo((int, int) x, int y) {
        X = x;
        Y = y;
    }

    // 这是一个解构方法
    public void Deconstruct(out (int, int) x, out int y) {
        x = X;
        y = Y;
    }
}

class Program {
    static void Main() {
        Console.WriteLine("This is C# Language");

        var foo = new Foo((1, 2), 3);

        // 利用解构方法解构Foo
        var (x, y) = foo;
        var ((a, b), _) = foo; // 使用_忽略不需要的值

        Console.WriteLine($"a = {a}, b = {b},  y = {y}");

        // 如果你想要重命名变量，可以这样做：
        var (j, i) = foo;
        Console.WriteLine($"i = {i}, j = ({j.Item1}, {j.Item2})");

        // 忽略x只取y
        var (_, yOnly) = foo;
        Console.WriteLine($"y = {yOnly}");
    }
}
/*Output
a = 1, b = 2,  y = 3
i = 3, j = (1, 2)
y = 3
*/

File: 08FlowControl/match/destructure_structures.rs
fn main() {
    struct Foo { x: (u32, u32), y: u32 }

    // 解构结构体的成员
    let foo = Foo { x: (1, 2), y: 3 };
    let Foo { x: (a, b), y } = foo;

    println!("a = {}, b = {},  y = {} ", a, b, y);

    // 可以解构结构体并重命名变量，成员顺序并不重要
	//Foo是结构体名字。
    let Foo { y: i, x: j } = foo;
    println!("i = {:?}, j = {:?}", i, j);

    // 也可以忽略某些变量，名字索引为主，哪怕是用的是y，它优先名字匹配/
    let Foo { y, .. } = foo;
    println!("y = {}", y);

    // 这将得到一个错误：模式中没有提及 `x` 字段
    let Foo { x, .. } = foo;
    println!("x = {:?}", x);
}
